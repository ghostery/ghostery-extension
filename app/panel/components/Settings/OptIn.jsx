/**
 * Opt-In Settings Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2020 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import PropTypes from 'prop-types';
import globals from '../../../../src/classes/Globals';

const { IS_CLIQZ } = globals;

const TOOLTIP_SVG_FILEPATH = '../../app/images/panel/icon-information-tooltip.svg';

/**
 * @class Implement Opt In subview as a React component.
 * The view opens from the left-side menu of the main Settings view.
 * It invites user to opt in for telemetry options and human web
 * @memberOf SettingsComponents
 */
const OptIn = ({ settingsData, toggleCheckbox }) => {
	const checkbox = (opt, name) => (
		<input
			type="checkbox"
			id={`settings-${opt}`}
			name={name}
			defaultChecked={settingsData[name]}
			onClick={toggleCheckbox}
		/>
	);

	const labelFor = (opt, text) => (
		<label htmlFor={`settings-${opt}`}>
			<span>{text}</span>
		</label>
	);

	const tooltipSVG = (text, dir) => (
		<div className={`s-tooltip-${dir}`} data-g-tooltip={text}>
			<img src={TOOLTIP_SVG_FILEPATH} className="s-question" />
		</div>
	);

	const option = (cbox, label, tooltip, id = '') => (
		<div className="s-option-group" id={id}>
			<div className="s-square-checkbox">
				{cbox}
				{label}
				{tooltip}
			</div>
		</div>
	);

	return (
		<div className="s-tabs-panel">
			<div className="row">
				<div className="columns">
					<h3>{t('settings_support_ghostery')}</h3>
					<h5>
						{t('settings_support_ghostery_by')}
						:
					</h5>
					{option(
						checkbox('share-usage', 'enable_metrics'),
						labelFor('share-usage', t('settings_share_usage')),
						tooltipSVG(t('settings_share_usage_tooltip'), 'down')
					)}
					{!IS_CLIQZ && option(
						checkbox('share-human-web', 'enable_human_web'),
						labelFor('share-human-web', t('settings_share_human_web')),
						tooltipSVG(t('settings_human_web_tooltip'), 'up'),
						'human-web-section'
					)}
					{option(
						checkbox('allow-abtests', 'enable_abtests'),
						labelFor('allow-abtests', t('settings_allow_abtests')),
						tooltipSVG(t('settings_abtests_tooltip'), 'up'),
						'abtests-section'
					)}
				</div>
			</div>
		</div>
	);
};

OptIn.propTypes = {
	toggleCheckbox: PropTypes.func.isRequired,
	settingsData: PropTypes.shape({
		enable_metrics: PropTypes.bool.isRequired,
		enable_human_web: PropTypes.bool.isRequired,
		enable_abtests: PropTypes.bool.isRequired,
	}).isRequired,
};

export default OptIn;
