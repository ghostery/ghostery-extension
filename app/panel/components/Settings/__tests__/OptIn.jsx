/**
 * OptIn Settings Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2020 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import OptIn from '../OptIn';

describe('app/panel/Settings/OptIn.jsx', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('OptIn is rendered correctly with falsy props', () => {
			const settingsData = {
				enable_metrics: false,
				enable_human_web: false,
			};

			const component = renderer.create(
				<OptIn
					settingsData={settingsData}
					toggleCheckbox={() => {}}
				/>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('OptIn is rendered correctly with truthy props', () => {
			const settingsData = {
				enable_metrics: true,
				enable_human_web: true,
			};

			const component = renderer.create(
				<OptIn
					settingsData={settingsData}
					toggleCheckbox={() => {}}
				/>
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Shallow snapshot tests rendered with Enzyme', () => {
		test('OptIn functions correctly', () => {
			const settingsData = {
				enable_metrics: true,
				enable_human_web: true,
			};
			const toggleCheckbox = jest.fn();

			const component = mount(
				<OptIn
					settingsData={settingsData}
					toggleCheckbox={toggleCheckbox}
				/>
			);

			expect(toggleCheckbox.mock.calls.length).toBe(0);
			component.find('#settings-share-usage').simulate('click');
			component.find('#settings-share-human-web').simulate('click');
			expect(toggleCheckbox.mock.calls.length).toBe(2);
		});
	});
});
