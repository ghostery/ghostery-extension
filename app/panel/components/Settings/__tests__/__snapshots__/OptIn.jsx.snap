// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/panel/Settings/OptIn.jsx Snapshot tests with react-test-renderer OptIn is rendered correctly with falsy props 1`] = `
<div
  className="s-tabs-panel"
>
  <div
    className="row"
  >
    <div
      className="columns"
    >
      <h3>
        settings_support_ghostery
      </h3>
      <h5>
        settings_support_ghostery_by
        :
      </h5>
      <div
        className="s-option-group"
        id=""
      >
        <div
          className="s-square-checkbox"
        >
          <input
            defaultChecked={false}
            id="settings-share-usage"
            name="enable_metrics"
            onClick={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="settings-share-usage"
          >
            <span>
              settings_share_usage
            </span>
          </label>
          <div
            className="s-tooltip-down"
            data-g-tooltip="settings_share_usage_tooltip"
          >
            <img
              className="s-question"
              src="../../app/images/panel/icon-information-tooltip.svg"
            />
          </div>
        </div>
      </div>
      <div
        className="s-option-group"
        id="human-web-section"
      >
        <div
          className="s-square-checkbox"
        >
          <input
            defaultChecked={false}
            id="settings-share-human-web"
            name="enable_human_web"
            onClick={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="settings-share-human-web"
          >
            <span>
              settings_share_human_web
            </span>
          </label>
          <div
            className="s-tooltip-up"
            data-g-tooltip="settings_human_web_tooltip"
          >
            <img
              className="s-question"
              src="../../app/images/panel/icon-information-tooltip.svg"
            />
          </div>
        </div>
      </div>
      <div
        className="s-option-group"
        id="abtests-section"
      >
        <div
          className="s-square-checkbox"
        >
          <input
            id="settings-allow-abtests"
            name="enable_abtests"
            onClick={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="settings-allow-abtests"
          >
            <span>
              settings_allow_abtests
            </span>
          </label>
          <div
            className="s-tooltip-up"
            data-g-tooltip="settings_abtests_tooltip"
          >
            <img
              className="s-question"
              src="../../app/images/panel/icon-information-tooltip.svg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`app/panel/Settings/OptIn.jsx Snapshot tests with react-test-renderer OptIn is rendered correctly with truthy props 1`] = `
<div
  className="s-tabs-panel"
>
  <div
    className="row"
  >
    <div
      className="columns"
    >
      <h3>
        settings_support_ghostery
      </h3>
      <h5>
        settings_support_ghostery_by
        :
      </h5>
      <div
        className="s-option-group"
        id=""
      >
        <div
          className="s-square-checkbox"
        >
          <input
            defaultChecked={true}
            id="settings-share-usage"
            name="enable_metrics"
            onClick={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="settings-share-usage"
          >
            <span>
              settings_share_usage
            </span>
          </label>
          <div
            className="s-tooltip-down"
            data-g-tooltip="settings_share_usage_tooltip"
          >
            <img
              className="s-question"
              src="../../app/images/panel/icon-information-tooltip.svg"
            />
          </div>
        </div>
      </div>
      <div
        className="s-option-group"
        id="human-web-section"
      >
        <div
          className="s-square-checkbox"
        >
          <input
            defaultChecked={true}
            id="settings-share-human-web"
            name="enable_human_web"
            onClick={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="settings-share-human-web"
          >
            <span>
              settings_share_human_web
            </span>
          </label>
          <div
            className="s-tooltip-up"
            data-g-tooltip="settings_human_web_tooltip"
          >
            <img
              className="s-question"
              src="../../app/images/panel/icon-information-tooltip.svg"
            />
          </div>
        </div>
      </div>
      <div
        className="s-option-group"
        id="abtests-section"
      >
        <div
          className="s-square-checkbox"
        >
          <input
            id="settings-allow-abtests"
            name="enable_abtests"
            onClick={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="settings-allow-abtests"
          >
            <span>
              settings_allow_abtests
            </span>
          </label>
          <div
            className="s-tooltip-up"
            data-g-tooltip="settings_abtests_tooltip"
          >
            <img
              className="s-question"
              src="../../app/images/panel/icon-information-tooltip.svg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
